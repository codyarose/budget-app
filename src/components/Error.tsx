import React, { FC } from 'react'

interface Props {
	error: boolean
	helperText: string
}

const Error: FC<Props> = ({ error, helperText }) => {
	return <>{error && <div>{helperText}</div>}</>
}

export default Error
