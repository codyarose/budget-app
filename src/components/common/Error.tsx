import React, { FC } from 'react'
import styled from 'styled-components'

interface Props {
	error: boolean
	helperText: string
}

const Error: FC<Props> = ({ error, helperText }) => {
	return <>{error && <StyledError>{helperText}</StyledError>}</>
}

export default Error

const StyledError = styled.div`
	font-style: italic;
`
