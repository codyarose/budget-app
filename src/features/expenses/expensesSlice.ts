import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { v4 as uuid } from 'uuid'
import { Moment } from 'moment'

import { RootState } from '../../app/store'
import moment from 'moment'

export interface Expense {
	id: string
	description: string
	note: string
	amount: number
	createdAt: Moment | number
}

export type ExpenseData = Omit<Expense, 'id'>

export const initialState: Expense[] = []

export const expensesSlice = createSlice({
	name: 'expenses',
	initialState,
	reducers: {
		addExpense: (state, action: PayloadAction<Partial<ExpenseData>>) => {
			const {
				description = '',
				note = '',
				amount = 0,
				createdAt = 0,
			} = action.payload
			state.push({
				id: uuid(),
				description,
				note,
				amount,
				createdAt,
			})
		},
		removeExpense: (state, action: PayloadAction<string>) => {
			const id = action.payload
			const index = state.findIndex((expense) => expense.id === id)
			if (index === -1) {
				return state
			} else {
				state.splice(index, 1)
			}
		},
		editExpense: (
			state,
			action: PayloadAction<{
				editedExpense: Partial<ExpenseData>
				id: string
			}>,
		) => {
			const { editedExpense, id } = action.payload
			return state.map((expense) => {
				if (expense.id === id) {
					return {
						...expense,
						...editedExpense,
					}
				} else {
					return expense
				}
			})
		},
	},
})

export const { addExpense, removeExpense, editExpense } = expensesSlice.actions

export const selectExpenses = (state: RootState): Expense[] => state.expenses

export const selectVisibleExpenses = (state: RootState): Expense[] => {
	const expenses = state.expenses
	const { startDate, endDate, text, sortBy } = state.filter

	return expenses
		.filter((expense) => {
			const createdAtMoment = moment(expense.createdAt)
			const momentStartDate = moment(startDate)
			const momentEndDate = moment(endDate)

			const startDateMatch = startDate
				? momentStartDate.isSameOrBefore(createdAtMoment, 'day')
				: true
			const endDateMatch = endDate
				? momentEndDate.isSameOrAfter(createdAtMoment, 'day')
				: true
			const textMatch = expense.description
				.toLowerCase()
				.includes(text.toLowerCase())

			return startDateMatch && endDateMatch && textMatch
		})
		.sort((a, b) => {
			if (sortBy === 'date') {
				return a.createdAt < b.createdAt ? 1 : -1
			} else if (sortBy === 'amount') {
				return a.amount < b.amount ? 1 : -1
			} else {
				return 1
			}
		})
}

export const selectTotal = (state: RootState): number => {
	const expenses = selectVisibleExpenses(state)

	if (!expenses.length) return 0

	return expenses.reduce((acc, expense) => {
		return acc + expense.amount
	}, 0)
}

export default expensesSlice.reducer
