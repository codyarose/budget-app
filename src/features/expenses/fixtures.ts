export default [
	{
		id: '123abc',
		description: 'Title',
		note: '',
		amount: 140500,
		createdAt: 1598763926001,
	},
	{
		id: '456def',
		description: 'Blah',
		note: 'This is a note',
		amount: 2650,
		createdAt: 1598763926009,
	},
	{
		id: '789ghi',
		description: 'Another one',
		note: 'Text goes here',
		amount: 340098,
		createdAt: 1598763926345,
	},
]
