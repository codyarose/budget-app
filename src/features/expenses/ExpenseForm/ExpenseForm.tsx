import React, { FC, useState, ChangeEvent, FormEvent } from 'react'
import styled from 'styled-components'
import moment, { Moment } from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/initialize'

import { ExpenseData } from '../expensesSlice'
import Error from '../../../components/Error'

interface Props {
	expense?: ExpenseData
	onSubmit: (arg0: ExpenseData) => void
}

interface FormState extends Omit<ExpenseData, 'amount'> {
	amount: string
}

const ExpenseForm: FC<Props> = ({ expense, onSubmit }) => {
	const [formState, setFormState] = useState<FormState>({
		description: expense?.description || '',
		amount: expense ? (expense?.amount / 100).toString() : '',
		note: expense?.note || '',
		createdAt: expense?.createdAt || moment().valueOf(),
	})

	const [error, setError] = useState('')

	const [calendarFocused, setCalendarFocused] = useState<boolean | null>(
		false,
	)

	const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
		const { name, value } = e.target
		setFormState({
			...formState,
			[name]: value,
		})
	}

	const handleAmountChange = (e: ChangeEvent<HTMLInputElement>) => {
		const amount = e.target.value
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			setFormState({ ...formState, amount })
		}
	}

	const handleSubmit = (e: FormEvent) => {
		e.preventDefault()
		const { description, amount } = formState
		if (!description || !amount) {
			setError('Please provide description and amount.')
		} else {
			setError('')
			const parsedAmount = parseFloat(amount) * 100
			onSubmit({
				...formState,
				amount: parsedAmount,
			})
		}
	}

	const handleDateChange = (createdAt: Moment | null) => {
		createdAt &&
			setFormState({ ...formState, createdAt: createdAt.valueOf() })
	}

	const handleFocusChange = (arg: { focused: boolean | null }) => {
		setCalendarFocused(arg.focused)
	}

	return (
		<div>
			<StyledForm onSubmit={handleSubmit}>
				<input
					type="text"
					placeholder="Description"
					name="description"
					id="expenseDescription"
					autoFocus
					value={formState.description}
					onChange={handleInputChange}
				/>
				<input
					type="text"
					placeholder="Amount"
					name="amount"
					id="expenseAmount"
					value={formState.amount || ''}
					onChange={handleAmountChange}
				/>
				<Error error={!!error} helperText={error} />
				<SingleDatePicker
					date={moment(formState.createdAt)}
					onDateChange={handleDateChange}
					focused={calendarFocused}
					onFocusChange={handleFocusChange}
					id="datePicker"
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
				<input
					type="text"
					name="note"
					id="expenseNote"
					placeholder="Add a not for your expense (optional)"
					value={formState.note}
					onChange={handleInputChange}
				/>
				<button type="submit">Add expense</button>
			</StyledForm>
		</div>
	)
}

export default ExpenseForm

const StyledForm = styled.form`
	display: flex;
	flex-flow: column;
`

StyledForm.displayName = 'StyledForm'
