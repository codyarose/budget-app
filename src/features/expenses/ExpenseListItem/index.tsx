import React, { FC } from 'react'
import { Link } from 'react-router-dom'

import { Expense } from '../expensesSlice'

const ExpenseListItem: FC<Expense> = ({
	description,
	amount,
	createdAt,
	note,
	id,
}) => {
	return (
		<div>
			<Link to={`/edit/${id}`}>
				<h4>{description}</h4>
			</Link>
			<p>
				{amount} - {createdAt}
			</p>
			{note && <p>{note}</p>}
		</div>
	)
}

export default ExpenseListItem
