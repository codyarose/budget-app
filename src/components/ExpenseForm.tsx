import React, { FC, useState, ChangeEvent, FormEvent } from 'react'
import styled from 'styled-components'
import moment, { Moment } from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import 'react-dates/initialize'

import { Expense } from '../features/expenses/expensesSlice'
import Error from './Error'

interface Props {
	expense?: Expense
	onSubmit: (arg0: Partial<Expense>) => void
}

interface FormState extends Omit<Expense, 'amount'> {
	amount: string
}

const ExpenseForm: FC<Props> = ({ expense, onSubmit }) => {
	const [formState, setFormState] = useState<Partial<FormState>>({
		description: expense?.description || '',
		amount: expense ? (expense?.amount / 100).toString() : '',
		note: expense?.note || '',
		createdAt: expense?.createdAt || moment().valueOf(),
	})

	const [error, setError] = useState('')

	const [calendarFocused, setCalendarFocused] = useState<boolean | null>(
		false,
	)

	const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
		const { name, value } = e.target
		setFormState({
			...formState,
			[name]: value,
		})
	}

	const handleAmountChange = (e: ChangeEvent<HTMLInputElement>) => {
		const amount = e.target.value
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			setFormState({ ...formState, amount })
		}
	}

	const handleSubmit = (e: FormEvent) => {
		e.preventDefault()
		const { description, amount } = formState
		if (!description || !amount) {
			setError('Please provide description and amount.')
		} else {
			setError('')
			const parsedAmount = parseFloat(amount) * 100
			onSubmit({
				...formState,
				amount: parsedAmount,
			})
		}
	}

	const handleDateChange = (createdAt: Moment | null) => {
		createdAt &&
			setFormState({ ...formState, createdAt: createdAt.valueOf() })
	}

	const handleFocusChange = (arg: { focused: boolean | null }) => {
		setCalendarFocused(arg.focused)
	}

	return (
		<div>
			<StyledForm onSubmit={handleSubmit}>
				<input
					type="text"
					placeholder="Description"
					name="description"
					autoFocus
					value={formState.description}
					onChange={handleInputChange}
				/>
				<input
					type="text"
					placeholder="Amount"
					name="amount"
					value={formState.amount || ''}
					onChange={handleAmountChange}
				/>
				<Error error={!!error} helperText={error} />
				<SingleDatePicker
					date={moment(formState.createdAt)}
					onDateChange={handleDateChange}
					focused={calendarFocused}
					onFocusChange={handleFocusChange}
					id="datePicker"
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
				<input
					type="text"
					name="note"
					placeholder="Add a not for your expense (optional)"
					value={formState.note}
					onChange={handleInputChange}
				/>
				<button type="submit">Add expense</button>
			</StyledForm>
		</div>
	)
}

export default ExpenseForm

const StyledForm = styled.form`
	display: flex;
	flex-flow: column;
`
