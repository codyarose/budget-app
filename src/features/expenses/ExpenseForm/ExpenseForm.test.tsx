import React from 'react'
import { shallow } from 'enzyme'

import ExpenseForm from './index'
import expenses from '../fixtures'
import Error from '../../../components/Error'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

jest.mock('moment', () => {
	const moment = jest.requireActual('moment')
	const mockMoment = (timestamp = 0) => {
		return moment(timestamp)
	}
	return mockMoment
})

describe('ExpenseForm', () => {
	it('should render correctly', () => {
		const props = { onSubmit: jest.fn() }
		const wrapper = shallow(<ExpenseForm {...props} />)

		expect(wrapper).toMatchSnapshot()
	})

	it('should render with expense data', () => {
		const props = { onSubmit: jest.fn(), expense: expenses[2] }
		const wrapper = shallow(<ExpenseForm {...props} />)

		expect(wrapper).toMatchSnapshot()
	})

	it('should render error for invalid form submission', () => {
		const props = { onSubmit: jest.fn() }
		const wrapper = shallow(<ExpenseForm {...props} />)
		const form = wrapper.find('StyledForm')
		form.simulate('submit', {
			preventDefault: () => null,
		})

		expect(wrapper.find(Error).props().helperText).toBe(
			'Please provide description and amount.',
		)
		expect(wrapper).toMatchSnapshot()
	})

	it('should set description on input change', () => {
		const props = { onSubmit: jest.fn() }
		const wrapper = shallow(<ExpenseForm {...props} />)
		const value = 'new description'
		wrapper.find('#expenseDescription').simulate('change', {
			target: { name: 'description', value },
		})

		expect(wrapper.find('#expenseDescription').props().value).toEqual(value)
	})

	it('should set note on input change', () => {
		const props = { onSubmit: jest.fn() }

		const wrapper = shallow(<ExpenseForm {...props} />)
		const value = 'new note'

		wrapper.find('#expenseNote').simulate('change', {
			target: { name: 'note', value },
		})
		expect(wrapper.find('#expenseNote').props().value).toEqual(value)
	})

	it('should set amount if valid input', () => {
		const props = { onSubmit: jest.fn() }
		const wrapper = shallow(<ExpenseForm {...props} />)
		const value = '13.69'

		wrapper.find('#expenseAmount').simulate('change', {
			target: { value },
		})
		expect(wrapper.find('#expenseAmount').props().value).toEqual(value)
	})

	it('should not set amount if invalid input', () => {
		const props = { onSubmit: jest.fn() }
		const wrapper = shallow(<ExpenseForm {...props} />)
		const value = '13.666'

		wrapper.find('#expenseAmount').simulate('change', {
			target: { value },
		})
		expect(wrapper.find('#expenseAmount').props().value).toEqual('')
	})

	it('should call onSubmit prop for valid form submission', () => {
		const onSubmitSpy = jest.fn()
		const props = { onSubmit: onSubmitSpy, expense: expenses[2] }
		const wrapper = shallow(<ExpenseForm {...props} />)
		const values = { ...expenses[2] }
		delete values.id

		wrapper.find('StyledForm').simulate('submit', {
			preventDefault: () => null,
		})

		expect(wrapper.find(Error).props().helperText).toBe('')
		expect(onSubmitSpy).toHaveBeenLastCalledWith(values)
	})

	it('should set new date on date change', () => {
		const props = { onSubmit: jest.fn() }
		const wrapper = shallow(<ExpenseForm {...props} />)
		const now = moment()
		wrapper.find(SingleDatePicker).prop('onDateChange')(now)

		expect(wrapper.find(SingleDatePicker).prop('date')).toEqual(now)
	})

	it('should set calendar focus on change', () => {
		const props = { onSubmit: jest.fn() }
		const wrapper = shallow(<ExpenseForm {...props} />)
		const focused = true
		wrapper.find(SingleDatePicker).prop('onFocusChange')({ focused })

		expect(wrapper.find(SingleDatePicker).prop('focused')).toEqual(focused)
	})
})
