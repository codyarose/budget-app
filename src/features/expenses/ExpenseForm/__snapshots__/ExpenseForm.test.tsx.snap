// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseForm should render correctly 1`] = `
<StyledForm
  onSubmit={[Function]}
>
  <styled.input
    autoFocus={true}
    id="expenseDescription"
    name="description"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <styled.input
    id="expenseAmount"
    name="amount"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <StyledSingleDatePicker
    date={"1970-01-01T00:00:00.000Z"}
    focused={false}
    id="datePicker"
    isOutsideRange={[Function]}
    numberOfMonths={1}
    onDateChange={[Function]}
    onFocusChange={[Function]}
  />
  <styled.input
    id="expenseNote"
    name="note"
    onChange={[Function]}
    placeholder="Add a note to your expense (optional)"
    type="text"
    value=""
  />
  <Error
    error={false}
    helperText=""
  />
  <Styled(Button)>
    Add
     expense
  </Styled(Button)>
</StyledForm>
`;

exports[`ExpenseForm should render error for invalid form submission 1`] = `
<StyledForm
  onSubmit={[Function]}
>
  <styled.input
    autoFocus={true}
    id="expenseDescription"
    name="description"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value=""
  />
  <styled.input
    id="expenseAmount"
    name="amount"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value=""
  />
  <StyledSingleDatePicker
    date={"1970-01-01T00:00:00.000Z"}
    focused={false}
    id="datePicker"
    isOutsideRange={[Function]}
    numberOfMonths={1}
    onDateChange={[Function]}
    onFocusChange={[Function]}
  />
  <styled.input
    id="expenseNote"
    name="note"
    onChange={[Function]}
    placeholder="Add a note to your expense (optional)"
    type="text"
    value=""
  />
  <Error
    error={true}
    helperText="Please provide description and amount."
  />
  <Styled(Button)>
    Add
     expense
  </Styled(Button)>
</StyledForm>
`;

exports[`ExpenseForm should render with expense data 1`] = `
<StyledForm
  onSubmit={[Function]}
>
  <styled.input
    autoFocus={true}
    id="expenseDescription"
    name="description"
    onChange={[Function]}
    placeholder="Description"
    type="text"
    value="Another one"
  />
  <styled.input
    id="expenseAmount"
    name="amount"
    onChange={[Function]}
    placeholder="Amount"
    type="text"
    value="3400.98"
  />
  <StyledSingleDatePicker
    date={"2020-08-30T05:05:26.345Z"}
    focused={false}
    id="datePicker"
    isOutsideRange={[Function]}
    numberOfMonths={1}
    onDateChange={[Function]}
    onFocusChange={[Function]}
  />
  <styled.input
    id="expenseNote"
    name="note"
    onChange={[Function]}
    placeholder="Add a note to your expense (optional)"
    type="text"
    value="Text goes here"
  />
  <Error
    error={false}
    helperText=""
  />
  <Styled(Button)>
    Update
     expense
  </Styled(Button)>
</StyledForm>
`;
